zrenderer/geometry/zrbvh/rbvhtree.cpp

```diff
@@ -0,0 +1,586 @@
+ *                     serkan.ergun@gmail.com
+ *
+ * This file is part of Z-Renderer <https://github.com/ZombieRendering/Z-Renderer>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <zrenderer/zcommon/mesh.h>
+#include <zrenderer/zcommon/utils.h>
+#include <zrenderer/geometry/zrbvh/rbvhtree.h>
+
+#include <zrenderer/geometry/zrbvh/shaders/fragRaster.glsl.h>
+#include <zrenderer/geometry/zrbvh/shaders/vertRaster.glsl.h>
+
+#define NOMINMAX
+#include <boost/filesystem/fstream.hpp>
+#include <oglplus/gl.hpp>
+#include <oglplus/all.hpp>
+#include <oglplus/bound/framebuffer.hpp>
+#include <oglplus/bound/renderbuffer.hpp>
+#include <FreeImage.h>
+
+namespace zrenderer
+{
+
+struct RBVHTriangle
+{
+    AlignedBox3f bound;
+    Vector3f normal;
+    float surfArea;
+};
+
+typedef std::vector<RBVHTriangle> RBVHTriangles;
+typedef std::shared_ptr<RBVHTriangles> RBVHTrianglesPtr;
+typedef std::vector<uint32_t> RBVHTriangleIndices;
+typedef std::shared_ptr<RBVHTriangleIndices> RBVHTriangleIndicesPtr;
+
+class RBVHNode
+{
+public:
+    RBVHNode( ConstMeshPtr mesh, MeshPtr& triMesh, float rho, float alpha,
+        float upperLimit )
+        : _triangles( new RBVHTriangles() )
+        , _triangleIndices( new RBVHTriangleIndices() )
+    {
+        const Vector3fs& vertices = mesh->getVertices();
+        const Vector3uis& faces = mesh->getFaces();
+
+        _triangles->reserve( faces.size() );
+        _triangleIndices->reserve( faces.size() );
+
+        uint32_t i = 0;
+        for( Vector3ui face : faces )
+        {
+            const Vector3f& v0 = vertices[ face.x() ];
+            const Vector3f& v1 = vertices[ face.y() ];
+            const Vector3f& v2 = vertices[ face.z() ];
+
+            Vector3f e1 = v1 - v0;
+            Vector3f e2 = v2 - v0;
+
+            _triangles->push_back( RBVHTriangle() );
+            _triangleIndices->push_back( i++ );
+            RBVHTriangle& triangle = _triangles->back();
+
+            triangle.bound.extend( v0 );
+            triangle.bound.extend( v1 );
+            triangle.bound.extend( v2 );
+            triangle.normal = e1.cross( e2 );
+            triangle.surfArea = 0.5f * triangle.normal.norm();
+            triangle.normal /= 2.0f * triangle.surfArea;
+        }
+
+        calcProperties( 0, (uint32_t)_triangles->size(), rho, alpha, upperLimit );
```

Comment:
calcProperties fonksiyonu o node için boundingBox, toplam yüzey alanı, yüzey alanı ile ağırlıklandırılmış normallerin toplamı gibi şeyleri hesaplıyor. split sırasındaki heuristiclerde kullanmak için. Nasıl bir isim önerirsin?